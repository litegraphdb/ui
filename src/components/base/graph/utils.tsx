import { NodeData } from '@/lib/graph/types';
import { defaultNodeColor, nodeLightColorByType, universalNodeColors } from './constant';

const maxLegends = 16;

export const getLegendsForNodes = (nodes: NodeData[]) => {
  const priorityLegends = new Map<string, { legend: string; color: string }>();
  const customLegends = new Map<string, { legend: string; color: string }>();

  // Separate legends into priority (nodeLightColorByType) and custom
  nodes.forEach((node) => {
    if (Object.keys(nodeLightColorByType).includes(node.type)) {
      priorityLegends.set(node.type, { legend: node.type, color: nodeLightColorByType[node.type] });
    } else {
      // Only add if not already present to avoid changing colors for same legend
      if (!customLegends.has(node.type)) {
        customLegends.set(node.type, {
          legend: node.type,
          color: universalNodeColors[customLegends.size % universalNodeColors.length],
        });
      }
    }
  });

  const priorityLegendsSize = priorityLegends.size;
  // If there are nodes with unknown types, add "Unknown" to priority legends
  const isMoreThanMaxLegends = customLegends.size + priorityLegendsSize > maxLegends;

  // Convert to arrays for easier manipulation
  const priorityArray = Array.from(priorityLegends.values());
  const customArray = Array.from(customLegends.values()).slice(
    0,
    maxLegends - (priorityArray.length + (isMoreThanMaxLegends ? 1 : 0))
  );

  // Combine priority legends first, then custom legends
  const legends = [
    ...priorityArray,
    ...customArray,
    ...(isMoreThanMaxLegends ? [{ legend: 'Unknown', color: defaultNodeColor }] : []),
  ];
  const mapLegends: Record<string, { legend: string; color: string }> = {};
  legends.forEach((legend) => {
    mapLegends[legend.legend] = legend;
  });
  return mapLegends;
};
