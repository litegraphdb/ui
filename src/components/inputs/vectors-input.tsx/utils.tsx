export const convertVectorsToAPIRecord = (
  vectors?: Array<{ Model: string; Dimensionality: number; Content: string; Vectors: number[] }>
) => {
  return (
    vectors?.map((vector: any) => ({
      ...vector,
      Dimensionality: Number(vector.Dimensionality),
      Vectors: vector.Vectors || [],
    })) || []
  );
};
