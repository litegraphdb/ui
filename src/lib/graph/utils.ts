import type { NodeData, Point } from './types';

export function isPointInNode(point: Point, node: NodeData, radius: number = 10): boolean {
  const dx = point.x - node.x;
  const dy = point.y - node.y;
  return dx * dx + dy * dy <= radius * radius;
}

export function getNodeAtPosition(nodes: NodeData[], point: Point): NodeData | null {
  return nodes.find((node) => isPointInNode(point, node)) || null;
}

export function transformToOptions<T extends { GUID: string; name?: string; Name?: string }>(
  data?: T[] | null,
  labelField: keyof T = 'name' // Field to use for label in options
) {
  return data
    ? data?.map((item) => ({
        value: item.GUID,
        label: (item[labelField] as string) || item.Name || item.GUID,
      }))
    : [];
}
