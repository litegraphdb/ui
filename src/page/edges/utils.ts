import { EdgeType } from '@/lib/store/edge/types';
import { NodeType } from '@/lib/store/node/types';

// Utility to find the name by GUID
export const getNodeNameByGUID = (guid: string, nodesList: NodeType[]) => {
  if (!guid) return 'N/A';
  const node = nodesList.find((n: NodeType) => n.GUID === guid);
  return node ? node.Name : guid;
};

export const transformEdgeDataForTable = (edgesList: EdgeType[], nodesList: NodeType[]) => {
  return (
    edgesList?.map((data: EdgeType) => {
      return {
        ...data,
        FromName: getNodeNameByGUID(data.From, nodesList),
        ToName: getNodeNameByGUID(data.To, nodesList),
        key: data.GUID,
      } as EdgeType;
    }) || []
  );
};
