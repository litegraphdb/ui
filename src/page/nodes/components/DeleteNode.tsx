import React, { Dispatch, SetStateAction } from 'react';
import toast from 'react-hot-toast';
import LitegraphModal from '@/components/base/modal/Modal';
import LitegraphParagraph from '@/components/base/typograpghy/Paragraph';
import LitegraphButton from '@/components/base/button/Button';
import { useAppDispatch } from '@/lib/store/hooks';
import { deleteNode } from '@/lib/store/node/actions';
import { useDeleteNodeById } from '@/lib/sdk/litegraph.service';
import { NodeType } from '@/lib/store/node/types';
import { useDeleteNodeMutation } from '@/lib/store/slice/slice';

type DeleteNodeProps = {
  title: string;
  paragraphText: string;
  isDeleteModelVisible: boolean;
  setIsDeleteModelVisible: Dispatch<SetStateAction<boolean>>;
  selectedNode: NodeType | null | undefined;
  setSelectedNode: Dispatch<SetStateAction<NodeType | null | undefined>>;
  onNodeDeleted?: () => Promise<void>;
};
const DeleteNode = ({
  title,
  paragraphText,
  isDeleteModelVisible,
  setIsDeleteModelVisible,
  selectedNode,
  setSelectedNode,
  onNodeDeleted,
}: DeleteNodeProps) => {
  const dispatch = useAppDispatch();

  const [deleteNodeById, { isLoading: isDeleteNodeLoading }] = useDeleteNodeMutation();

  const handleDeleteNode = async () => {
    if (selectedNode) {
      const res = await deleteNodeById({
        graphId: selectedNode.GraphGUID,
        nodeId: selectedNode.GUID,
      });
      if (res) {
        dispatch(deleteNode({ GUID: selectedNode.GUID }));
        toast.success('Delete Node successfully');
        setIsDeleteModelVisible(false);
        setSelectedNode(null);
        onNodeDeleted && onNodeDeleted();
      }
    }
  };
  return (
    <LitegraphModal
      title={title}
      centered
      open={isDeleteModelVisible}
      onCancel={() => setIsDeleteModelVisible(false)}
      footer={
        <LitegraphButton type="primary" onClick={handleDeleteNode} loading={isDeleteNodeLoading}>
          Confirm
        </LitegraphButton>
      }
    >
      <LitegraphParagraph>{paragraphText}</LitegraphParagraph>
    </LitegraphModal>
  );
};

export default DeleteNode;
