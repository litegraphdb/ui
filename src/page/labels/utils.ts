import { NodeType } from '@/types/types';
import { getNodeNameByGUID } from '../edges/utils';

import { getEdgeNameByGUID } from '../tags/utils';
import { LabelMetadata } from 'litegraphdb/dist/types/types';
import { EdgeType } from '@/types/types';
import { LabelMetadataForTable } from './types';

export const transformLabelsDataForTable = (
  labelsList: LabelMetadata[],
  nodesList: NodeType[],
  edgesList: EdgeType[]
) => {
  return (
    labelsList?.map((data: LabelMetadata) => {
      return {
        ...data,
        NodeName: getNodeNameByGUID(data.NodeGUID, nodesList),
        EdgeName: getEdgeNameByGUID(data.EdgeGUID, edgesList),
        key: data.GUID,
      } as LabelMetadataForTable;
    }) || []
  );
};

export const getNodeAndEdgeGUIDsByLabelList = (labelsList: LabelMetadata[]) => {
  const nodeGUIDs = labelsList.map((label) => label.NodeGUID);
  const edgeGUIDs = labelsList.map((label) => label.EdgeGUID);
  return {
    nodeGUIDs: Array.from(new Set(nodeGUIDs)),
    edgeGUIDs: Array.from(new Set(edgeGUIDs)),
  };
};
