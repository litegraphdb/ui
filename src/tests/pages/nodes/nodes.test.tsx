import '@testing-library/jest-dom';
import React from 'react';
import { screen, fireEvent, waitFor } from '@testing-library/react';
import NodePage from '../../../page/nodes/NodePage';
import { createMockInitialState } from '../../store/mockStore';
import { mockGraphGUID, mockNodeData } from '../mockData';
import { setupServer } from 'msw/node';
import { handlers } from './handler';
import { commonHandlers } from '@/tests/handler';
import { renderWithRedux } from '@/tests/store/utils';
import AddEditNode from '@/page/nodes/components/AddEditNode';
import DeleteNode from '@/page/nodes/components/DeleteNode';

const server = setupServer(...handlers, ...commonHandlers);

describe('NodesPage', () => {
  beforeAll(() => server.listen());
  afterEach(() => {
    server.resetHandlers();
  });
  afterAll(() => server.close());

  it('renders the nodes page', async () => {
    const initialState = createMockInitialState();
    const { container } = renderWithRedux(<NodePage />, initialState, undefined, true);

    await waitFor(() => {
      expect(screen.getByText(/nodes/i)).toBeVisible();
      expect(screen.getByRole('button', { name: /create node/i })).toBeVisible();
    });
    expect(container).toMatchSnapshot('initial table state');
  });

  it('should display Create Node button', () => {
    const initialState = createMockInitialState();
    const { container } = renderWithRedux(<NodePage />, initialState, undefined, true);

    const createButton = screen.getByRole('button', { name: /create node/i });
    expect(createButton).toBeVisible();
    expect(createButton).toMatchSnapshot();
  });

  it('should create a node and should be visible in the table', async () => {
    const initialState = createMockInitialState();
    const { container } = renderWithRedux(<AddEditNode isAddEditNodeVisible={true} setIsAddEditNodeVisible={() => {}} node={null} selectedGraph={mockGraphGUID} />, initialState, undefined, true);

    const modal = await screen.findByTestId('add-edit-node-modal');
    expect(modal).toBeInTheDocument();

    await waitFor(() => {
      expect(screen.queryByText('Loading...')).not.toBeInTheDocument();
    }, { timeout: 10000 });

    const nameInput = screen.getByPlaceholderText(/enter node name/i);
    fireEvent.change(nameInput, { target: { value: mockNodeData[0].Name } });

    await waitFor(() => {
      expect(nameInput.value).toBe(mockNodeData[0].Name);
    });

    const createButton = screen.getByRole('button', { name: /create/i });
    fireEvent.click(createButton);

    expect(container).toMatchSnapshot('final table state after creation');
  });

  it('should update a node successfully', async () => {
    const initialState = createMockInitialState();
    const { container } = renderWithRedux(<AddEditNode isAddEditNodeVisible={true} setIsAddEditNodeVisible={() => {}} node={mockNodeData[0]} selectedGraph={mockGraphGUID} readonly={false}/>, initialState, undefined, true);

    const modal = await screen.findByTestId('add-edit-node-modal');
    expect(modal).toBeInTheDocument();

    await waitFor(() => {
      expect(screen.queryByText('Loading...')).not.toBeInTheDocument();
    }, { timeout: 10000 });

    const nameInput = screen.getByPlaceholderText(/enter node name/i);
    fireEvent.change(nameInput, { target: { value: 'My updated test node' } });

    await waitFor(() => {
      expect(nameInput.value).toBe('My updated test node');
    });

    const createButton = screen.getByRole('button', { name: /update/i });
    fireEvent.click(createButton);

    expect(container).toMatchSnapshot('final table state after update');
  }); 

  it('should delete a node successfully', async () => {
    const initialState = createMockInitialState();
    const { container } = renderWithRedux(<DeleteNode isDeleteModelVisible={true} setIsDeleteModelVisible={() => {}} selectedNode={mockNodeData[0]} setSelectedNode={() => {}} title={`Are you sure you want to delete "${mockNodeData[0].Name}" node?`} paragraphText={'This action will delete node.'}/>, initialState, undefined, true);

    const modal = await screen.findByTestId('delete-node-modal');
    expect(modal).toBeInTheDocument();

    const deleteButton = screen.getByRole('button', { name: /confirm/i });
    fireEvent.click(deleteButton);

    expect(container).toMatchSnapshot('final table state after deletion');
  }); 
}); 