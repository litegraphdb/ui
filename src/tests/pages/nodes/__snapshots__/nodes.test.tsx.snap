// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NodesPage renders the nodes page: initial table state 1`] = `
<div>
  <main
    class="ant-layout-content pageContainer css-dev-only-do-not-override-98ntnt"
    id="nodes"
  >
    <div
      class="pageTitleContainer ant-flex css-dev-only-do-not-override-98ntnt ant-flex-wrap-wrap ant-flex-align-center ant-flex-justify-space-between ant-flex-gap-small"
    >
      <span
        class="ant-typography css-dev-only-do-not-override-98ntnt"
        data-testid="heading"
        style="font-weight: 600; font-size: 16px;"
      >
        <div
          class="ant-flex css-dev-only-do-not-override-98ntnt ant-flex-align-center"
          style="gap: 10px;"
        >
          <span
            class="ant-typography css-dev-only-do-not-override-98ntnt"
          >
            Nodes
          </span>
          <span
            aria-label="search"
            class="anticon anticon-search cursor-pointer"
            role="img"
            tabindex="-1"
          >
            <svg
              aria-hidden="true"
              data-icon="search"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
              />
            </svg>
          </span>
          <span
            aria-label="loading"
            class="anticon anticon-loading anticon-spin loading-icon"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="loading"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="0 0 1024 1024"
              width="1em"
            >
              <path
                d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
              />
            </svg>
          </span>
        </div>
      </span>
      <button
        class="ant-btn css-dev-only-do-not-override-98ntnt ant-btn-link"
        style="font-weight: 500;"
        type="button"
      >
        <span
          class="ant-btn-icon"
        >
          <span
            aria-label="plus-square"
            class="anticon anticon-plus-square"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="plus-square"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M328 544h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8z"
              />
              <path
                d="M880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656z"
              />
            </svg>
          </span>
        </span>
        <span>
          Create Node
        </span>
      </button>
    </div>
    <div
      class="pageContent"
    >
      <div
        class="mb-sm ant-flex css-dev-only-do-not-override-98ntnt ant-flex-align-center ant-flex-justify-space-between"
        style="margin-top: -10px; gap: 20px;"
      />
      <div
        class="ant-table-wrapper css-dev-only-do-not-override-98ntnt"
      >
        <div
          class="ant-spin-nested-loading css-dev-only-do-not-override-98ntnt"
        >
          <div>
            <div
              aria-busy="true"
              aria-live="polite"
              class="ant-spin ant-spin-spinning css-dev-only-do-not-override-98ntnt"
            >
              <span
                class="ant-spin-dot-holder"
              >
                <span
                  class="ant-spin-dot ant-spin-dot-spin"
                >
                  <i
                    class="ant-spin-dot-item"
                  />
                  <i
                    class="ant-spin-dot-item"
                  />
                  <i
                    class="ant-spin-dot-item"
                  />
                  <i
                    class="ant-spin-dot-item"
                  />
                </span>
              </span>
            </div>
          </div>
          <div
            class="ant-spin-container ant-spin-blur"
          >
            <div
              class="ant-table ant-table-empty css-dev-only-do-not-override-98ntnt"
            >
              <div
                class="ant-table-container"
              >
                <div
                  class="ant-table-content"
                >
                  <table
                    style="table-layout: auto;"
                  >
                    <colgroup>
                      <col
                        style="width: 250px;"
                      />
                      <col
                        style="width: 350px;"
                      />
                      <col
                        style="width: 150px;"
                      />
                      <col
                        style="width: 250px;"
                      />
                      <col
                        style="width: 250px;"
                      />
                      <col
                        style="width: 250px;"
                      />
                    </colgroup>
                    <thead
                      class="ant-table-thead"
                    >
                      <tr>
                        <th
                          aria-label="Name"
                          class="ant-table-cell ant-table-column-has-sorters react-resizable"
                          scope="col"
                          tabindex="0"
                        >
                          <div
                            class="ant-table-filter-column"
                          >
                            <span
                              class="ant-table-column-title"
                            >
                              <div
                                class="ant-table-column-sorters"
                              >
                                <span
                                  class="ant-table-column-title"
                                >
                                  Name
                                </span>
                                <span
                                  class="ant-table-column-sorter ant-table-column-sorter-full"
                                >
                                  <span
                                    aria-hidden="true"
                                    class="ant-table-column-sorter-inner"
                                  >
                                    <span
                                      aria-label="caret-up"
                                      class="anticon anticon-caret-up ant-table-column-sorter-up"
                                      role="img"
                                    >
                                      <svg
                                        aria-hidden="true"
                                        data-icon="caret-up"
                                        fill="currentColor"
                                        focusable="false"
                                        height="1em"
                                        viewBox="0 0 1024 1024"
                                        width="1em"
                                      >
                                        <path
                                          d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                        />
                                      </svg>
                                    </span>
                                    <span
                                      aria-label="caret-down"
                                      class="anticon anticon-caret-down ant-table-column-sorter-down"
                                      role="img"
                                    >
                                      <svg
                                        aria-hidden="true"
                                        data-icon="caret-down"
                                        fill="currentColor"
                                        focusable="false"
                                        height="1em"
                                        viewBox="0 0 1024 1024"
                                        width="1em"
                                      >
                                        <path
                                          d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                        />
                                      </svg>
                                    </span>
                                  </span>
                                </span>
                              </div>
                            </span>
                            <span
                              class="ant-dropdown-trigger ant-table-filter-trigger"
                              role="button"
                              tabindex="-1"
                            >
                              <span
                                aria-label="filter"
                                class="anticon anticon-filter"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="filter"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="64 64 896 896"
                                  width="1em"
                                >
                                  <path
                                    d="M349 838c0 17.7 14.2 32 31.8 32h262.4c17.6 0 31.8-14.3 31.8-32V642H349v196zm531.1-684H143.9c-24.5 0-39.8 26.7-27.5 48l221.3 376h348.8l221.3-376c12.1-21.3-3.2-48-27.7-48z"
                                  />
                                </svg>
                              </span>
                            </span>
                          </div>
                          <span
                            class="react-resizable-handle"
                          />
                        </th>
                        <th
                          class="ant-table-cell react-resizable"
                          scope="col"
                        >
                          <div
                            class="ant-table-filter-column"
                          >
                            <span
                              class="ant-table-column-title"
                            >
                              GUID
                            </span>
                            <span
                              class="ant-dropdown-trigger ant-table-filter-trigger"
                              role="button"
                              tabindex="-1"
                            >
                              <span
                                aria-label="filter"
                                class="anticon anticon-filter"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="filter"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="64 64 896 896"
                                  width="1em"
                                >
                                  <path
                                    d="M349 838c0 17.7 14.2 32 31.8 32h262.4c17.6 0 31.8-14.3 31.8-32V642H349v196zm531.1-684H143.9c-24.5 0-39.8 26.7-27.5 48l221.3 376h348.8l221.3-376c12.1-21.3-3.2-48-27.7-48z"
                                  />
                                </svg>
                              </span>
                            </span>
                          </div>
                          <span
                            class="react-resizable-handle"
                          />
                        </th>
                        <th
                          class="ant-table-cell react-resizable"
                          scope="col"
                        >
                          <div
                            class="ant-table-filter-column"
                          >
                            <span
                              class="ant-table-column-title"
                            >
                              Labels
                            </span>
                            <span
                              class="ant-dropdown-trigger ant-table-filter-trigger"
                              role="button"
                              tabindex="-1"
                            >
                              <span
                                aria-label="filter"
                                class="anticon anticon-filter"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="filter"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="64 64 896 896"
                                  width="1em"
                                >
                                  <path
                                    d="M349 838c0 17.7 14.2 32 31.8 32h262.4c17.6 0 31.8-14.3 31.8-32V642H349v196zm531.1-684H143.9c-24.5 0-39.8 26.7-27.5 48l221.3 376h348.8l221.3-376c12.1-21.3-3.2-48-27.7-48z"
                                  />
                                </svg>
                              </span>
                            </span>
                          </div>
                          <span
                            class="react-resizable-handle"
                          />
                        </th>
                        <th
                          class="ant-table-cell react-resizable"
                          scope="col"
                        >
                          <div
                            class="ant-table-filter-column"
                          >
                            <span
                              class="ant-table-column-title"
                            >
                              Tags
                            </span>
                            <span
                              class="ant-dropdown-trigger ant-table-filter-trigger"
                              role="button"
                              tabindex="-1"
                            >
                              <span
                                aria-label="filter"
                                class="anticon anticon-filter"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="filter"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="64 64 896 896"
                                  width="1em"
                                >
                                  <path
                                    d="M349 838c0 17.7 14.2 32 31.8 32h262.4c17.6 0 31.8-14.3 31.8-32V642H349v196zm531.1-684H143.9c-24.5 0-39.8 26.7-27.5 48l221.3 376h348.8l221.3-376c12.1-21.3-3.2-48-27.7-48z"
                                  />
                                </svg>
                              </span>
                            </span>
                          </div>
                          <span
                            class="react-resizable-handle"
                          />
                        </th>
                        <th
                          class="ant-table-cell react-resizable"
                          scope="col"
                        >
                          Vectors
                          <span
                            class="react-resizable-handle"
                          />
                        </th>
                        <th
                          aria-label="Created UTC"
                          class="ant-table-cell ant-table-column-has-sorters react-resizable"
                          scope="col"
                          tabindex="0"
                        >
                          <div
                            class="ant-table-column-sorters"
                          >
                            <span
                              class="ant-table-column-title"
                            >
                              Created UTC
                            </span>
                            <span
                              class="ant-table-column-sorter ant-table-column-sorter-full"
                            >
                              <span
                                aria-hidden="true"
                                class="ant-table-column-sorter-inner"
                              >
                                <span
                                  aria-label="caret-up"
                                  class="anticon anticon-caret-up ant-table-column-sorter-up"
                                  role="img"
                                >
                                  <svg
                                    aria-hidden="true"
                                    data-icon="caret-up"
                                    fill="currentColor"
                                    focusable="false"
                                    height="1em"
                                    viewBox="0 0 1024 1024"
                                    width="1em"
                                  >
                                    <path
                                      d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                    />
                                  </svg>
                                </span>
                                <span
                                  aria-label="caret-down"
                                  class="anticon anticon-caret-down ant-table-column-sorter-down"
                                  role="img"
                                >
                                  <svg
                                    aria-hidden="true"
                                    data-icon="caret-down"
                                    fill="currentColor"
                                    focusable="false"
                                    height="1em"
                                    viewBox="0 0 1024 1024"
                                    width="1em"
                                  >
                                    <path
                                      d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                    />
                                  </svg>
                                </span>
                              </span>
                            </span>
                          </div>
                          <span
                            class="react-resizable-handle"
                          />
                        </th>
                        <th
                          class="ant-table-cell"
                          scope="col"
                        >
                          Actions
                        </th>
                      </tr>
                    </thead>
                    <tbody
                      class="ant-table-tbody"
                    >
                      <tr
                        class="ant-table-placeholder"
                      >
                        <td
                          class="ant-table-cell"
                          colspan="7"
                        >
                          <div
                            class="css-dev-only-do-not-override-98ntnt ant-empty ant-empty-normal"
                          >
                            <div
                              class="ant-empty-image"
                            >
                              <svg
                                height="41"
                                viewBox="0 0 64 41"
                                width="64"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <title>
                                  Simple Empty
                                </title>
                                <g
                                  fill="none"
                                  fill-rule="evenodd"
                                  transform="translate(0 1)"
                                >
                                  <ellipse
                                    cx="32"
                                    cy="33"
                                    fill="#f5f5f5"
                                    rx="32"
                                    ry="7"
                                  />
                                  <g
                                    fill-rule="nonzero"
                                    stroke="#d9d9d9"
                                  >
                                    <path
                                      d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                                    />
                                    <path
                                      d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                                      fill="#fafafa"
                                    />
                                  </g>
                                </g>
                              </svg>
                            </div>
                            <div
                              class="ant-empty-description"
                            >
                              No data
                            </div>
                          </div>
                        </td>
                      </tr>
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </main>
</div>
`;

exports[`NodesPage should create a node and should be visible in the table: final table state after creation 1`] = `<div />`;

exports[`NodesPage should delete a node successfully: final table state after deletion 1`] = `<div />`;

exports[`NodesPage should display Create Node button 1`] = `
<button
  class="ant-btn css-dev-only-do-not-override-98ntnt ant-btn-link"
  style="font-weight: 500;"
  type="button"
>
  <span
    class="ant-btn-icon"
  >
    <span
      aria-label="plus-square"
      class="anticon anticon-plus-square"
      role="img"
    >
      <svg
        aria-hidden="true"
        data-icon="plus-square"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M328 544h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8z"
        />
        <path
          d="M880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656z"
        />
      </svg>
    </span>
  </span>
  <span>
    Create Node
  </span>
</button>
`;

exports[`NodesPage should update a node successfully: final table state after update 1`] = `<div />`;
